hphp/runtime/vm/jit/irlower-ret.cpp

```diff
@@ -95,7 +95,16 @@ void prepare_return_regs(Vout& v, SSATmp* retVal, Vloc retLoc,
     auto const extended = v.makeReg();
     auto const result = v.makeReg();
 
-    v << movzbq{type, extended};
+    switch (arch()) {
+      case Arch::X64: // fall through
+      case Arch::PPC64:
+        v << movzbq{type, extended};
+        break;
+      case Arch::ARM:
+        // For ARM64 we need to clear the bits 8..63 from the type value.
+        v << andqi{0xFF, type, extended, v.makeReg()};
+        break;
```

Comment:
Perhaps we can use the following:

    // Explicitly enforce the invariant that bits 31..8 are zero for Vreg8's
    // until movzbq{} and similar instructions are optimized to simple mov{}
    // in a more careful and selective way.
    v << movzbq{type, extended};
    v << andqi{0xFF, extended, extended, v.makeReg()};